module/zfs/zap.c

```diff
@@ -920,21 +921,17 @@ fzap_add_cd(zap_name_t *zn,
 	} else if (err == EAGAIN) {
 		err = zap_expand_leaf(zn, l, tag, tx, &l);
 		zap = zn->zn_zap;	/* zap_expand_leaf() may change zap */
-		if (err == 0) {
+		if (err == 0)
 			goto retry;
-		} else if (err == ENOSPC) {
-			/*
-			 * If we failed to expand the leaf, then bailout
-			 * as there is no point trying
-			 * zap_put_leaf_maybe_grow_ptrtbl().
-			 */
-			return (err);
-		}
 	}
 
 out:
-	if (zap != NULL)
-		zap_put_leaf_maybe_grow_ptrtbl(zn, l, tag, tx);
+	if (l != NULL) {
+		if (err == ENOSPC)
+			zap_put_leaf(l);
```

Comment:
(not a review, I'm trying to understand this)

it gets here because it was `zap_grow_ptrtbl` which returned `ENOSPC`; when can that happen? a) the ZAP is way too large (2^46 entries, right?) and b) on transition from 2^28 to 2^48 `zap_hashbits`, that is when upgrading from micro to fat zap? Is there any other case that I'm missing why ENOSPC can happen here?

I mean, I'm trying to understand which separate two leaves does the `zap_put_leaf` get called for and whether that could cause additional trouble - this is the second occasion, the first is on line 713.

Reply:
I've primarily just refactored already present code.  I was actually thinking about actual out of space, but it seems not checked at this point, and you seem right that hash size limit may be the primary (and pretty unlikely) factor.

zap_expand_leaf() puts one leaf inside, while another (where the entry is going to be inserted) returns to the caller.  So it is the leaf that gets new entry is checked for coming overflow and possible hash table grow.  I.e. I see no problems, unless you specify better.